### ISSUE TITLE
Allow handling of written numbers

### ISSUE THREAD

**Issue by ErykPiasecki07** [ErykPiasecki07/VerbCalc#30]
Make 'two' appear as '2' for example.

**Comment by LucaMellini** [comment]
hi, I would like to work on that issue please !

**Comment by ErykPiasecki07** [comment]
Welcome in the project!

Thanks for your willingness to contribute, I am assigning the issue right away!

**Comment by syedatif4118** [comment]
Hi! I am just a beginner in Python, is there any basic issue where I can contribute to the project?

**Comment by ErykPiasecki07** [comment]
Hi! This issue is a pretty good starting point, you might get it if there is still no response from @LucaMellini, as he seems not to give any signs of life here. If he however will give some quick heads-up in this conversation, #34 seems to be a pretty good starting point.

**Comment by ErykPiasecki07** [comment]
Just let me know if #34 seems right for you!

**Comment by LucaMellini** [comment]
hi ! I'm sorry for the giving no sign of life but I had a lot to do recently and didnt have time to dive into it yet. 
However this week will be way more calm so I will be able to give it a real start !
Sorry for the disappointment though

**Comment by syedatif4118** [comment]
I will check out #34 and get back to you asap

**Comment by LucaMellini** [comment]
hey @ErykPiasecki07 !
I  finished my task in local (I've done lots of unit tests and my solution seems to work pretty well).
I still got some questions before integrating it into the project. It's mainly about how deep I have to go :  

1. Do I have to take hyphen into account? I'm asking cause the hyphen could be confused with a minus, right? And I dont really know how to manage it.  Like if someone wants to write "twenty-one", how can we know if he tries to write "21" or to make a substraction? Do we allow people to make substraction using the minus symbol as a string?

2. How do you plan to manage the "and" word? In my solution, writing "one thousand and one" will output "1001" but it also could be an operation with a logical AND.

thanks in advance !

**Comment by ErykPiasecki07** [comment]
Hi @LucaMellini !
First off:

> Sorry for the disappointment

No worries, there's no disappointment, everyone has personal life, right? It's great to hear back from you! In regards to later questions:

1. No need for that, generally the library is meant for 'human speech' -> 'mathematical operations' conversion, not the mix of both. We would just expect user not to input any non-alphanumeric characters, btw. that sounds like a great issue, simple regex would do - I will make one after I finish this reply.

2. Despite unmeasured importance to the field of mathematics the logical connectives have, I don't think we have to go that deep. If the library is to be reliant on human speech, I don't think many people would actually be willing to ask such complex questions via their speech, let alone if they use it like a voice-assistant for quickly calculating their bills for instance. However, as new issues are trying to bring modularity via plugins to the whole thing, one day we might take a bite of that problem again to make logical connectives possible.

3. Regarding the details like which exact function or file it should be in, I cannot really tell, it's really about your vision! What I can tell, is that it should be somehow embedded to the 'translation' rather than 'dispatching' or 'calculation'. Regarding second part of that question, library already does that! If you worked fully locally, your code be quite a bit behind. If I were you, I would try to clone the newest version and try to fit your existing code into that, since resolving conflicts in merge later will be a tedious process otherwise. 

> Sorry for thoses noob questions but I'm very new in the world of Open Source

Those were really valid and important questions, after all: "better to ask the way than to go astray" as they say, right? 

Thanks once more for all the contribution!

**Comment by LucaMellini** [comment]
Alright thanks for your answers @ErykPiasecki07 !
I'll work on that and try to create the PR before the end of the week !

**Comment by LucaMellini** [comment]
Hey @ErykPiasecki07,
I was wondering if it was possible to import external modules into the project?
(the "re" module for example)
thank you!

**Comment by ErykPiasecki07** [comment]
I hope you will be able to figure something out, of course usage of regex is allowed (remember to add it to both requirements.txt and requirements-dev.txt)

**Comment by LucaMellini** [comment]
Alright ! Everything seems working now !
There is just one last thing I have to fix (concerning the invalid expressions like "2 2" or "+ -").
Every other tests are green !  
I'll also clean up the code and comment it to make it as clear as possible.
Almost finished !

**Comment by ErykPiasecki07** [comment]
Great to hear about your progress!

Regarding your question: the code's way of defining **valid** expression, is that after doing all the calculations there's only one number remaining AND that at any point there wasn't attempt of performing operation on any symbol.

**Comment by ErykPiasecki07** [comment]
Following my previous comment, have in mind that all the checks I wrote above are just results of code itself, there's no any 'check validity' function. Code doesn't know whether the calculation is valid until it tries to solve it. In best scenario it realizes that the operation it's trying to perform is invalid (for ex. '2 + -') and in worst case after all the operations there is more than one number remaining with no operations to perform.

**Comment by LucaMellini** [comment]
Ok I understand.
The problem is that it is now something I have to check in the translation. Otherwise the translation wont succeed correctly. 
I have to do a check concerning the scales of the elements I receive or something like this.
At the moment, if I receive the sentence "two two" it will output "4" because I do a calculation between every written elements I receive in order to take in charge big numbers. Though it seems to perfectly work for "normal" numbers but not for invalid ones. I.E. "two three four plus five" will output "9 + 5" and so on.. 
It might be a little hard to deal with but I'll try to.

**Comment by LucaMellini** [comment]
Now that I'm thinking about it, the initial issue which is the handling of written numbers, is done and works well ! Maybe I'll create the  PR knowing that there is still that little bug? (for another issue maybe?) 
I dont know if I'll have time to keep improving this project in the following weeks. 
Don't hesitate to tag me if you have questions for the next issues that imply my code !
I'll be glad to help more !
Thank  you for your help and I hope what I brought will be useful to the project !

**Comment by ErykPiasecki07** [comment]
Hi! Sorry I wasn't available yesterday, but yeah, incorrectness in handling invalid statements is not exactly this issue's problem and needs its' own solution. I will check the pull request but it will be merged only after the solution to the invalid statements is found, because as for now your upgrade while very impressive would also make library less stable. Thanks for your contribution!

**Comment by ErykPiasecki07** [comment]
And quick disclaimer: don't worry if the review takes me a few days, but I'm having quite a madhouse in my private life right now, I promise I will get back to it whenever I am ready!

**Comment by LucaMellini** [comment]
> incorrectness in handling invalid statements is not exactly this issue's problem and needs its' own solution.

I'll try to keep working on it as much as I can with the time I have ! It kinda triggers me not having my job completely done..





> And quick disclaimer: don't worry if the review takes me a few days, but I'm having quite a madhouse in my private life right now, I promise I will get back to it whenever I am ready!

No worries for that, keep me informed whenever you want !

**Comment by ErykPiasecki07** [comment]
I have created separate issue for that, I will assign you if you feel like it's a job for you!

**Comment by ErykPiasecki07** [comment]
But for that to happen you have to comment anything in that issue, just Github things ;(

**Comment by LucaMellini** [comment]
> But for that to happen you have to comment anything in that issue, just Github things ;(

can you assign multiple contributors to an issue? I will comment it then !

**Comment by LucaMellini** [comment]
By the way, while doing my pull request, I found out that there is still a library called "word2number" that would have fixed this current issue. Have you heard about it? Would you consider using it? (Just asking by pure curiosity)

**Comment by ErykPiasecki07** [comment]
> can you assign multiple contributors to an issue? I will comment it then !

I think so, we will see!


> By the way, while doing my pull request, I found out that there is still a library called "word2number" that would have fixed this current issue. Have you heard about it? Would you consider using it?

I have never heard about it! Have you tried it? Is it stable? If so, I think it would be a way to go, especially if it fixes the issue! If you believe it might work, do some tests on it and get back with results!

> I was also asking myself : How did the idea of this project come to your mind?

It struck me as I was making a project for Intel® AI4Youth Polish (my home country) Edition. Part of this project was a virtual assistant, like Siri or Alexa. We implemented free version of Wolfram Alpha API as a proof of concept, but I thought to myself - there isn't anything like this open source! It would be easily embeddable to websites if one wants that, it would be easily embeddable to assistants and it would be easy to wrap with GUI as a standalone or even possibly a plugin to some other app!

Let me know when you play some with word2number! I'm really curious about the results!

**Comment by ErykPiasecki07** [comment]
Welcome back @LucaMellini!
How is it going with word2number and stuff?

**Comment by ErykPiasecki07** [comment]
Unassigning the issue since it went stale
